--- ./src/win/util.c	2013-04-10 13:43:07.000000000 -0400
+++ ./src/win/util.c	2013-05-12 01:04:16.955664000 -0400
@@ -31,7 +31,22 @@
 #include "uv.h"
 #include "internal.h"
 
-#include <Winsock2.h>
+#ifdef __MINGW32__
+    // MINGW64 previous to version 0.2.7 and 0.1.7 did not properly pull in
+    // dependencies needed for iphlpapi.h leading to compiler errors such as
+    // ras.h:988:3: error: unknown type name 'IsolationState'
+    // see: http://sourceforge.net/mailarchive/message.php?msg_id=29793138
+    #include <_mingw.h>
+    #if !defined(__MINGW64_VERSION_MAJOR)
+    #elif __MINGW64_VERSION_MAJOR > 2
+    #elif __MINGW64_VERSION_MINOR > 6
+    #else
+        #include <naptypes.h>
+        #include <ntddndis.h>
+    #endif
+#endif
+
+#include <winsock2.h>
 #include <iphlpapi.h>
 #include <psapi.h>
 #include <tlhelp32.h>

